File: sdk/typescript/src/webhooks.ts
import type { WebhookEvent } from "./models.js";
import { hmacBase64Url, sha256Digest, timingSafeEqual } from "./signing.js";

const encoder = new TextEncoder();
export interface VerifiedWebhook { event: WebhookEvent; eventId: string; keyId: string; timestamp: number; bodyDigest: string }
export type WebhookSecretResolver = (keyId: string) => string | undefined | Promise<string | undefined>;
export type InboxResult = "processed" | "duplicate" | "conflict";
export interface WebhookInbox<Transaction = unknown> {
  process(eventId: string, bodyDigest: string, handler: (transaction: Transaction) => Promise<void>): Promise<InboxResult>;
}
export class WebhookVerificationError extends Error { override name = "WebhookVerificationError"; }

export async function verifyWebhook(options: {
  rawBody: Uint8Array; signatureHeader: string; contentDigest: string; resolveSecret: WebhookSecretResolver;
  now?: number; toleranceSeconds?: number;
}): Promise<VerifiedWebhook> {
  const parts = Object.fromEntries(options.signatureHeader.split(",").map((part) => part.trim().split(/=(.*)/s).slice(0, 2)));
  const timestamp = Number(parts.t); const keyId = parts.key; const eventId = parts.event; const provided = parts.v1;
  if (!Number.isInteger(timestamp) || !keyId || !eventId || !provided) throw new WebhookVerificationError("invalid webhook signature header");
  const now = options.now ?? Math.floor(Date.now() / 1000);
  if (Math.abs(now - timestamp) > (options.toleranceSeconds ?? 300)) throw new WebhookVerificationError("webhook timestamp outside tolerance");
  const digest = await sha256Digest(options.rawBody);
  if (!timingSafeEqual(digest, options.contentDigest)) throw new WebhookVerificationError("webhook content digest mismatch");
  const secret = await options.resolveSecret(keyId);
  if (!secret) throw new WebhookVerificationError("unknown webhook key");
  const prefix = encoder.encode(`${eventId}.${timestamp}.`);
  const signingInput = new Uint8Array(prefix.length + options.rawBody.length); signingInput.set(prefix); signingInput.set(options.rawBody, prefix.length);
  const expected = await hmacBase64Url(secret, signingInput);
  if (!timingSafeEqual(expected, provided)) throw new WebhookVerificationError("webhook signature mismatch");
  let event: WebhookEvent;
  try { event = JSON.parse(new TextDecoder().decode(options.rawBody)) as WebhookEvent; } catch { throw new WebhookVerificationError("invalid webhook JSON"); }
  if (event.event_id !== eventId || event.schema_version !== "1") throw new WebhookVerificationError("webhook envelope mismatch");
  return { event, eventId, keyId, timestamp, bodyDigest: digest };
}
export function acknowledgement(eventId: string): { acknowledged_event_id: string } { return { acknowledged_event_id: eventId }; }

File: sdk/go/webhooks.go
package merchantplatform

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strconv"
	"strings"
	"time"
)

type VerifiedWebhook struct {
	Event                      WebhookEvent
	EventID, KeyID, BodyDigest string
	Timestamp                  int64
}
type SecretResolver func(keyID string) ([]byte, bool)
type InboxResult string

const (
	InboxProcessed InboxResult = "processed"
	InboxDuplicate InboxResult = "duplicate"
	InboxConflict  InboxResult = "conflict"
)

type WebhookInbox interface {
	Process(ctx context.Context, eventID, bodyDigest string, handler func(transaction any) error) (InboxResult, error)
}

func VerifyWebhook(rawBody []byte, signatureHeader, contentDigest string, resolve SecretResolver, now time.Time, tolerance time.Duration) (VerifiedWebhook, error) {
	parts := map[string]string{}
	for _, item := range strings.Split(signatureHeader, ",") {
		pair := strings.SplitN(strings.TrimSpace(item), "=", 2)
		if len(pair) == 2 {
			parts[pair[0]] = pair[1]
		}
	}
	timestamp, err := strconv.ParseInt(parts["t"], 10, 64)
	if err != nil || parts["key"] == "" || parts["event"] == "" || parts["v1"] == "" {
		return VerifiedWebhook{}, errors.New("invalid webhook signature header")
	}
	delta := now.Sub(time.Unix(timestamp, 0))
	if delta < 0 {
		delta = -delta
	}
	if delta > tolerance {
		return VerifiedWebhook{}, errors.New("webhook timestamp outside tolerance")
	}
	digest := sha256.Sum256(rawBody)
	digestHeader := "sha-256=:" + base64.StdEncoding.EncodeToString(digest[:]) + ":"
	if !hmac.Equal([]byte(digestHeader), []byte(contentDigest)) {
		return VerifiedWebhook{}, errors.New("webhook content digest mismatch")
	}
	secret, found := resolve(parts["key"])
	if !found {
		return VerifiedWebhook{}, errors.New("unknown webhook key")
	}
	mac := hmac.New(sha256.New, secret)
	_, _ = mac.Write(append([]byte(parts["event"]+"."+parts["t"]+"."), rawBody...))
	expected := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
	if !hmac.Equal([]byte(expected), []byte(parts["v1"])) {
		return VerifiedWebhook{}, errors.New("webhook signature mismatch")
	}
	var event WebhookEvent
	if err := json.Unmarshal(rawBody, &event); err != nil {
		return VerifiedWebhook{}, errors.New("invalid webhook JSON")
	}
	if event.EventID != parts["event"] || event.SchemaVersion != "1" {
		return VerifiedWebhook{}, errors.New("webhook envelope mismatch")
	}
	return VerifiedWebhook{event, event.EventID, parts["key"], digestHeader, timestamp}, nil
}
func Acknowledgement(eventID string) map[string]string {
	return map[string]string{"acknowledged_event_id": eventID}
}
